# Machine Learning Seminar
### Chapter06. Decision tree
<br>
19.08.28  


JaeEun Yoo


<hr>

## Decision Trees

* Like SVMs, Decision Trees are versatile Machine Learning algorithms that can perform both classification and regression tasks, and even multioutput tasks.
* Decision Trees are also the fundamental components of Random Forests (see Chapter 7), which are among the most powerful Machine Learning algorithms available today.
* In this chapter we will start by discussing how to train, visualize, and make predictions with Decision Trees.
* Then we will go through the CART training algorithm used by Scikit-Learn, and we will discuss how to regularize trees and use them for regression tasks. 
* Finally, we will discuss some of the limitations of Decision Trees.



<hr>

## Training and Visualizing a Decision Tree
To understand Decision Trees, let’s just build one and take a look at how it makes predictions.  
The following code trains a DecisionTreeClassifier on the iris dataset (see Chapter 4):

### Training

In [1]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier

In [2]:
iris = load_iris()
X = iris.data[:, 2:] # petal length and width
y = iris.target

In [3]:
print(iris)

{'data': array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
     

In [4]:
print(iris.keys())

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])


In [5]:
X

array([[1.4, 0.2],
       [1.4, 0.2],
       [1.3, 0.2],
       [1.5, 0.2],
       [1.4, 0.2],
       [1.7, 0.4],
       [1.4, 0.3],
       [1.5, 0.2],
       [1.4, 0.2],
       [1.5, 0.1],
       [1.5, 0.2],
       [1.6, 0.2],
       [1.4, 0.1],
       [1.1, 0.1],
       [1.2, 0.2],
       [1.5, 0.4],
       [1.3, 0.4],
       [1.4, 0.3],
       [1.7, 0.3],
       [1.5, 0.3],
       [1.7, 0.2],
       [1.5, 0.4],
       [1. , 0.2],
       [1.7, 0.5],
       [1.9, 0.2],
       [1.6, 0.2],
       [1.6, 0.4],
       [1.5, 0.2],
       [1.4, 0.2],
       [1.6, 0.2],
       [1.6, 0.2],
       [1.5, 0.4],
       [1.5, 0.1],
       [1.4, 0.2],
       [1.5, 0.2],
       [1.2, 0.2],
       [1.3, 0.2],
       [1.4, 0.1],
       [1.3, 0.2],
       [1.5, 0.2],
       [1.3, 0.3],
       [1.3, 0.3],
       [1.3, 0.2],
       [1.6, 0.6],
       [1.9, 0.4],
       [1.4, 0.3],
       [1.6, 0.2],
       [1.4, 0.2],
       [1.5, 0.2],
       [1.4, 0.2],
       [4.7, 1.4],
       [4.5, 1.5],
       [4.9,

In [6]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [7]:
tree_clf = DecisionTreeClassifier(max_depth=2)
tree_clf.fit(X, y)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=2,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

### Visualize

In [8]:
from sklearn.tree import export_graphviz

export_graphviz(tree_clf,
                out_file="iris_tree.dot",
                feature_names=iris.feature_names[2:],
                class_names=iris.target_names,
                rounded=True,
                filled=True)

<hr>

## Making Predictions

![Figure 6-1](./img/iris_tree.png)



* Suppose you find an iris flower and you want to classify it.
* You start at the **root node** (depth 0, at the top).
* If it is, then you move down to the root’s left child node (depth 1, left). In this case, it is a **leaf node** (i.e., it does not have any children nodes)
* Now suppose you find another flower, but this time the petal length is greater than 2.45 cm.
* It asks another question: is the petal width smaller than 1.75 cm? 
* If it is, then your flower is most likely an Iris- Versicolor (depth 2, left). 
* If not, it is likely an Iris-Virginica (depth 2, right).
* One of the many qualities of Decision Trees is that they require very little data preparation. 
* In particular, they don’t require feature scaling or centering at all.


### attribute

* samples : how many training instances it applies to
* value : how many training instances of each class this node applies to
* gini : impurity
    - A node is “pure” (gini=0) if all training instances it applies to belong to the same class.


![Figure 6-2](./img/06_01.PNG)



* For example, the depth-2 left node has a gini score equal to 1 – (0/54)<sup>2</sup>
    – (49/54)<sup>2</sup> – (5/54)<sup>2</sup> ≈ 0.168.
* p<sub>i,k</sub> is the ratio of class k instances among the training instances in the ith node.

![Figure 6-1](./img/iris_tree.png)



![Figure 6-2](./img/06_02.PNG)



* Figure 6-2 shows this Decision Tree’s decision boundaries. 
* The thick vertical line represents the decision boundary of the root node (depth 0): petal length = 2.45 cm. 
* Since the left area is pure (only Iris-Setosa), it cannot be split any further. 
* However, the right area is impure, so the depth-1 right node splits it at petal width = 1.75 cm (represented by the dashed line). 
* Since max_depth was set to 2, the Decision Tree stops right there. 
* However, if you set max_depth to 3, then the two depth-2 nodes would each add another decision boundary (represented by the dotted lines).

> MODEL INTERPRETATION: WHITE BOX VERSUS BLACK BOX

* As you can see Decision Trees are fairly intuitive and their decisions are easy to interpret. Such models are often called **white box models.**
* In contrast, as we will see, Random Forests or neural networks are generally considered **black box models.**
* They make great predictions, and you can easily check the calculations that they performed to make these predictions; nevertheless, it is usually hard to explain in simple terms why the predictions were made. 
    - For example, if a neural network says that a particular person appears on a picture, it is hard to know what actually contributed to this prediction: did the model recognize that person’s eyes? Her mouth? Her nose? Her shoes? Or even the couch that she was sitting on?
* Conversely, Decision Trees provide nice and simple classification rules that can even be applied manually if need be (e.g., for flower classification).



<br>
<hr>

## Estimating Class Probabilites

* A Decision Tree can also estimate the probability that an instance belongs to a particular class k :
    * first it traverses the tree to find the leaf node for this instance, and then it returns the ratio of training instances of class k in this node. 
* For example, suppose you have found a flower whose petals are 5cm long and 1.5 cm wide.
    * The corresponding leaf node is the depth-2 left node, so the Decision Tree should output the following probabilities: 0% for Iris-Setosa (0/54), 90.7% for Iris-Versicolor (49/54), and 9.3% for Iris-Virginica (5/54). 
* And of course if you ask it to predict the class, it should output Iris-Versicolor (class 1) since it has the highest probability. 

![Figure 6-1](./img/iris_tree.png)



In [9]:
tree_clf.predict_proba([[5, 1.5]])

array([[0.        , 0.90740741, 0.09259259]])

In [10]:
tree_clf.predict([[5, 1.5]])

array([1])

<hr>

## The CART Training Algorithm

* Scikit-Learn uses the **Classification And Regression Tree (CART) algorithm** to train Decision Trees (also called “growing” trees). 
* The idea is really quite simple: the algorithm first splits the training set in **two subsets** using a *single feature k* and a *threshold t<sub>k</sub>* (e.g., “petal length ≤ 2.45 cm”). 
* How does it choose k and tk? 
    - It searches for the pair (k, tk) that produces the purest subsets (weighted by their size). 
* The cost function that the algorithm tries to minimize is given by Equation 6-2.

![](./img/06_03.PNG)



* Once it has successfully split the training set in two, it splits the subsets using the same logic, then the sub-subsets and so on, recursively. 
* It stops recursing once it reaches the **maximum depth** (defined by the max_depth hyperparameter), or if it cannot find a split that will reduce **impurity**. 
* A few other hyperparameters (described in a moment) control additional stopping conditions (min_samples_split, min_samples_leaf, min_weight_fraction_leaf, and max_leaf_nodes).

>#### WARNING

As you can see, the CART algorithm is a **greedy algorithm**:  
it greedily searches for an optimum split at the top level, then
repeats the process at each level. It does not check whether or not the split will lead to the lowest possible impurity several
levels down.  
**A greedy algorithm often produces a reasonably good solution, but it is not guaranteed to be the optimal solution.**

<hr>

## Computational Complexity

![](./img/06_04.PNG)



* Making predictions requires traversing the Decision Tree from the root to a leaf. 
* Decision Trees are generally approximately balanced, so traversing the Decision Tree requires going through roughly **O(log<sub>2</sub>(m))** nodes.

> log<sub>2</sub> is the binary logarithm. It is equal to log<sub>2</sub>(m) = log(m) / log(2).  

![](./img/06_05.PNG)



* Since each node only requires checking the value of one feature, the overall prediction complexity is just O(log<sub>2</sub>(m)), independent of the number of features. 
* So predictions are very fast, even when dealing with large training sets.
* However, **the training algorithm compares all features (or less if max_features is set) on all samples at each node.**
* This results in a training complexity of **O(n × m log(m)).** 
* For small training sets (less than a few thousand instances), Scikit-Learn can speed up training by presorting the data (set presort=True), but this slows down training considerably for larger training sets.

<hr>

## Gini Impurity or Entropy?


* By default, the Gini impurity measure is used, but you can select the **entropy impurity** measure instead by setting the *criterion* hyperparameter to *"entropy"*. 
* The concept of entropy originated in thermodynamics as a measure of molecular disorder: entropy approaches zero when molecules are still and well ordered. 
* It later spread to a wide variety of domains, including Shannon’s information theory, where it measures the average information content of a message:
    * entropy is zero when all messages are identical. 
* In Machine Learning, it is frequently used as an impurity measure: a set’s entropy is zero when it contains instances of only one class.


![](./img/06_06.PNG)



* Equation 6-3 shows the definition of the entropy of the ith node.

* The truth is, most of the time it does not make a big difference: they lead to similar trees. 
* Gini impurity is slightly faster to compute, so it is a good default. 
* However, when they differ, Gini impurity tends to isolate the most frequent class in its own branch of the tree, while entropy tends to produce slightly more balanced trees.

<hr>

## Regularization Hyperparameters

* Decision Trees make very few assumptions about the training data (as opposed to linear models, which obviously assume that the data is linear, for example). 
* If left unconstrained, the tree structure will adapt itself to the training data, fitting it very closely, and most likely **overfitting** it. 
* Such a model is often called a **nonparametric model**, not because it does not have any parameters (it often has a lot) but because the number of parameters is not determined prior to training, so the model structure is free to stick closely to the data. 
* In contrast, a **parametric model** such as a linear model has a predetermined number of parameters, so its degree of freedom is limited, reducing the risk of overfitting (but increasing the risk of underfitting).

* To avoid overfitting the training data, you need to **restrict the Decision Tree’s freedom** during training.
    * As you know by now, this is called **regularization**. 
* The regularization hyperparameters depend on the algorithm used, but generally you can at least restrict the **maximum depth of the Decision Tree**. 
* In Scikit-Learn, this is controlled by the *max_depth* hyperparameter (the default value is None, which means unlimited). 
* Reducing max_depth will regularize the model and thus reduce the risk of overfitting.

* The DecisionTreeClassifier class has a few other parameters that similarly restrict the shape of the Decision Tree: 
    * min_samples_split (the minimum number of samples a node must have before it can be split)
    * min_samples_leaf (the minimum number of samples a leaf node must have)
    * min_weight_fraction_leaf (same as min_samples_leaf but expressed as a fraction of the total number of weighted instances)
    * max_leaf_nodes (maximum number of leaf nodes)
    * max_features (maximum number of features that are evaluated for splitting at each node).
* **Increasing min_ *hyperparameters or reducing max_* hyperparameters will regularize the model.**

>#### NOTE

Other algorithms work by first training the Decision Tree without restrictions, then pruning (deleting) unnecessary nodes.  
A node whose children are all leaf nodes is considered unnecessary if the purity improvement it provides is not *statistically significant*. 
Standard statistical tests, such as the χ2 test, are used to estimate the probability that the improvement is purely
the result of chance (which is called the *null hypothesis*). If this probability, called the p-value, is higher than a given
threshold (typically 5%, controlled by a hyperparameter), then the node is considered unnecessary and its children are
deleted. The pruning continues until all unnecessary nodes have been pruned.


![](./img/06_07.PNG)



* Figure 6-3 shows two Decision Trees trained on the moons dataset (introduced in Chapter 5). 
* On the left, the Decision Tree is trained with the default hyperparameters (i.e., no restrictions)
* On the right the Decision Tree is trained with *min_samples_leaf=4*. 
* It is quite obvious that the model on **the left is overfitting**, and **the model on the right will probably generalize better.**

<hr>

## Regression

Decision Trees are also capable of performing regression tasks. Let’s build a regression tree using Scikit-Learn’s *DecisionTreeRegressor* class, training it on a noisy quadratic dataset with *max_depth=2*:

In [11]:
from sklearn.tree import DecisionTreeRegressor
tree_reg = DecisionTreeRegressor(max_depth=2)
tree_reg.fit(X, y)

DecisionTreeRegressor(criterion='mse', max_depth=2, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=None, splitter='best')


![](./img/06_08.PNG)



* This tree looks very similar to the classification tree you built earlier. 
* The main difference is that instead of predicting a class in each node, it **predicts a value.** 
* For example, suppose you want to make a prediction for a new instance with x<sub>1</sub> = 0.6. 
    * You traverse the tree starting at the root, and you eventually reach the leaf node that predicts **value=0.1106**. 
* This prediction is simply the average target value of the 110 training instances associated to this leaf node. 
* This prediction results in a Mean Squared Error (MSE) equal to 0.0151 over these 110 instances.



![](./img/06_09.PNG)



* This model’s predictions are represented on the left of Figure 6-5. 
    * If you set *max_depth=3*, you get the predictions represented on the right. 
* Notice how the predicted value for each region is always the average target value of the instances in that region. 
* The algorithm splits each region in a way that makes most training instances as close as possible to that predicted value.

* The CART algorithm works mostly the same way as earlier, except that instead of trying to split the training set in a way that minimizes impurity, it now tries to split the training set in a way that **minimizes the MSE.** 
* Equation 6-4 shows the cost function that the algorithm tries to minimize.


![](./img/06_10.PNG)




![](./img/06_11.PNG)



* Just like for classification tasks, Decision Trees are prone to overfitting when dealing with regression tasks. 
* Without any regularization (i.e., using the default hyperparameters), you get the predictions on the left of Figure 6-6. 
    * It is obviously overfitting the training set very badly. 
* Just setting *min_samples_leaf=10* results in a much more reasonable model, represented on the right of Figure 6-6.

<hr>

## Instability

* However they do have a few limitations. 
* First, as you may have noticed, Decision Trees love orthogonal decision boundaries (all splits are perpendicular to an axis), which makes them sensitive to training set rotation. 
    * For example, Figure 6-7 shows a simple linearly separable dataset: on the left, a Decision Tree can split it easily, while on the right, after the dataset is rotated by 45°, the decision boundary looks unnecessarily convoluted. 
    * Although both Decision Trees fit the training set perfectly, it is very likely that the model on the right will not generalize well. 
    * One way to limit this problem is to use PCA (see Chapter 8), which often results in a better orientation of the training data.


![](./img/06_12.PNG)



* More generally, the main issue with Decision Trees is that they are very sensitive to small variations in the training data. 
    * For example, if you just remove the widest Iris-Versicolor from the iris training set (the one with petals 4.8 cm long and 1.8 cm wide) and train a new Decision Tree, you may get the model represented in Figure 6-8.


![](./img/06_13.PNG)



As you can see, it looks very different from the previous Decision
Tree (Figure 6-2).


![](./img/06_02.PNG)



* Actually, since the training algorithm used by Scikit-Learn is stochastic you may get very different models even on the same training data (unless you set the random_state hyperparameter). 
* Random Forests can limit this instability by averaging predictions over many trees, as we will see in the next chapter.